### ISSUE TITLE
Watch with resourceVersion="" can take down control plane

### ISSUE THREAD

**Issue by serathius** [kubernetes/kubernetes#123448]
### What happened?

I was investigating a customer case where the Kubernetes control plane was unavailable for prolonged time due to pod Watch being broken. I was suspecting an issue like https://github.com/etcd-io/etcd/issues/15402. The only clues I had was large sized of pods (>50KBs), watch cache being stale (`rage(apiserver_watch_cache_events_dispatched_total[1m])` equal zero) and a lot of errors `Fast watcher, slow processing` and occasional `prevKV=nil`.

Due to lack of proper watch instrumentation in kube-apiserver, I was only able to reproduce a similar situation on my local cluster. What I found is that even in large clusters with thousands of watches, misguided clients can still open a watch with `resourceVersion=""` that will directly read from etcd and if left inactive, can take down the whole control plane.

In my testing I have observed that slow/inactive watch on throughput heavy resource opened directly to etcd can cause:
* Huge memory leak on both kube-apiserver and etcd. This is caused by direct etcd watches having an infinite buffer and never kicking clients. 
* Overload of the apiserver-etcd watch stream, causing starvation of other watches, including a watch opened by watch cache. Starved watches experience a `prevKV=nil` error, causing them to break. If the watch opened by the watch cache is broken enough times it can cause termination of all watches and any new watch will be immediately closed. Short watches can lead to the whole APF being exhausted, as watches take the same amount of seats no matter how long they persist.

This goes against user expectation, where a large correctly running cluster with large pod objects can be taken down by single misguided client.

### What did you expect to happen?

* Inactive/slow watchers are kicked to prevent infinite memory increase. This already happens for watches opened to watch cache.
* It should not be possible for direct watches to etcd to cause termination of watch cache watches. Connection for watch cache should be separated to avoid cross pollution
* When watch cache is enabled, users should not be able to create direct watches to etcd. We need https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md for Watch.
* Window of available events be limited not by compaction window but by PrevKV availability
* Kube-apiserver should provide a metrics allowing to trace state of watch requests, also separate watches to etcd and watch cache.

Some of the issues listed here require improvements on etcd side. I will create separate issue in etcd repo after repeating the investigation for etcd alone.

### How can we reproduce it (as minimally and precisely as possible)?

My testing showed limit of 400MB/s of watch throughput. 

This was achieved by:
* 100 watches with resourceVersion="" to a single resource, each consuming 1 event per second
* 100 write QPS of 50KB objects to a single resource

To reproduce prevKV=nil, before my workstation run out of memory I also reduced the compaction window to 5s. 

Code used
```
package main

import (
	"context"
	"fmt"
	"math/rand"
	"path/filepath"
	"sync"
	"sync/atomic"
	"time"

	"k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/tools/clientcmd"
	"k8s.io/client-go/util/homedir"
)

var objectSize = 50000
var objectCount = 10
var timeBetweenRequests = time.Millisecond * 100
var maxQPS float32 = 100

var initResourceVersion = ""
var watchers = 100
var watcherSleep = time.Second * 1

func main() {
	config, err := clientcmd.BuildConfigFromFlags("", filepath.Join(homedir.HomeDir(), ".kube", "config"))
	if err != nil {
		panic(err.Error())
	}
	config.QPS = maxQPS
	config.Burst = 100

	// create the clientset
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		panic(err.Error())
	}

	var wg sync.WaitGroup
	var events atomic.Int64
	startWriters(clientset, &wg)
	startWatchers(clientset, &wg, &events)
	start := time.Now()
	events.Store(0)

	fmt.Printf("Start %s\n", start)
	for range time.Tick(time.Second) {
		fmt.Printf("%s events: %d\n", time.Since(start), events.Load())
	}
	wg.Wait()
}

func startWriters(clientset kubernetes.Interface, wg *sync.WaitGroup) {
	for i := 0; i < objectCount; i++ {
		name := fmt.Sprintf("%d", i)
		wg.Add(1)
		go func() {
			defer wg.Done()
			clientset.CoreV1().ConfigMaps("default").Delete(context.TODO(), name, metav1.DeleteOptions{})
			_, err := clientset.CoreV1().ConfigMaps("default").Create(context.TODO(), randomConfigmap(name), metav1.CreateOptions{})
			if err != nil {
				panic(err)
			}
			time.Sleep(timeBetweenRequests)
			for {
				_, err := clientset.CoreV1().ConfigMaps("default").Update(context.TODO(), randomConfigmap(name), metav1.UpdateOptions{})
				if err != nil {
					panic(err)
				}
				time.Sleep(timeBetweenRequests)
			}
		}()
	}
}
func startWatchers(clientset kubernetes.Interface, wg *sync.WaitGroup, events *atomic.Int64) {
	for i := 0; i < watchers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for {
				watch, err := clientset.CoreV1().ConfigMaps("default").Watch(context.TODO(), metav1.ListOptions{ResourceVersion: initResourceVersion})
				if err != nil {
					panic(err)
				}
				for event := range watch.ResultChan() {
					switch event.Object.(type) {
					case *v1.ConfigMap:
						events.Add(1)
					case *metav1.Status:
					default:
						fmt.Printf("Event, type: %s, obj: %+v\n", event.Type, event.Object)
					}
					time.Sleep(watcherSleep)
				}
			}
		}()
	}
}

func randomConfigmap(name string) *v1.ConfigMap {
	return &v1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name: name,
		},
		Immutable: nil,
		Data: map[string]string{
			"random": RandString(objectSize),
		},
		BinaryData: nil,
	}
}

const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

func RandString(l int) string {
	s := make([]byte, l)
	for i := 0; i < l; i++ {
		s[i] = chars[rand.Intn(len(chars))]
	}
	return string(s)
}
```

### Kubernetes version

1.29.2

### Cloud provider

KIND

### OS version

n/a

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by serathius** [comment]
Expect this is related to https://github.com/kubernetes/kubernetes/issues/123072

**Comment by serathius** [comment]
Question to @wojtek-t @jpbetz @deads2k 

Want to propose `ConsistentWatchFromCache`, that will follow the same principles as https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md. Do we need a separate KEP?

**Comment by dims** [comment]
+1 to pick the best existing KEP and add this case to it!

**Comment by dims** [comment]
@serathius i believe there was a C++ based client that was part of the mix in https://github.com/kubernetes/kubernetes/issues/123072 (believe enabling that component triggered the issue)

left pointers here https://github.com/kubernetes/kubernetes/issues/123072#issuecomment-1959729121

**Comment by MadhavJivrajani** [comment]
Would I be correct in understanding that if `rv=""` would lead to such a scenario, then any of the conditions here being true would lead to a similar situation?
https://github.com/kubernetes/kubernetes/blob/ef19539cdc3980be2637a33b9ab01dae5da71e08/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L735

**Comment by mengqiy** [comment]
In https://github.com/kubernetes/kubernetes/issues/123072, we saw the following:
- etcd memory kept increasing during the incident
- there were clients making large number of watch request w/o resourceVersion e.g. `/api/v1/watch/pods?fieldSelector=status.phase!=Failed,status.phase!=Unknown,status.phase!=Succeeded,spec.nodeName=ip-10-32-88-156.ec2.internal&pretty=false`
- `apiserver_watch_cache_events_received_total{resource="pods"}` diverged between 2 apiserver instances. Same for `apiserver_watch_cache_events_dispatched_total{resource="pods"}`
- `prevKV=nil` are seen during the incident.

**Comment by wojtek-t** [comment]
> Also, if I'm reading everything correctly, if we take [KEP-3157: Watch/List](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md) and remove the "list" part, it's very close to a "consistent watch from cache".

Yes - there is a plan to address that part there - we should just extend that KEP, the code change is pretty minimal (in fact I wanted this to happen anyway).

**Comment by shyamjvs** [comment]
So from @mengqiy repro above, our observation is that when there is a large number of watches going to etcd for a resource type that has high churn - it's leading to watch events being lost. Next @chaochn47 is going to confirm if all these watches are reusing the same underlying client connection between API server -> etcd. Based on this our next steps would be understanding if the event loss is happening due to some connection-level limitation or server-side bottleneck.

**Comment by serathius** [comment]
For issue of

>It should not be possible for direct watches to etcd to cause termination of watch cache watches. Connection for watch cache should be separated to avoid cross pollution

I have confirmed that the watch starvation can be resolved if we separate watches into separate RPCs. Etcd client doesn't provide a direct access to managing streams, however client creates a create RPC for unique set of grpc metadata. A separate streams can be forced by adding unique grpc metadata in the watch context.

 It is important to note that having separate RPC streams impacts progress notifications needed for some features like ConsistentListFromCache. For example RequestProgressNotification will be sent to the watch stream with the same grpc metadata, meaning that we should be careful not to add metadata to watch opened by watch cache so it can receive progress notifications.
 
Options:
* [recommended] Separate just watch opened by watch cache from all other. This should prevent watch cache being starved, while not adding a lot of overhead. This seems like make most sense short term.
* Make each watch separate, this would prevent any 2 watches to starve each other. Need more testing to confirm the performance impact. This opens the question of "why did etcd implement watch by multiplexing multiple watch calls on single watch grpc?". Don't know the answer yet, however it seems like a bad idea to implement it's own stream multiplexing over a protocol that already provides it.

Draft implementation of second option https://github.com/kubernetes/kubernetes/pull/123483

**Comment by serathius** [comment]
Did some testing performance of running each watch in separate RPC. Found that:
* It requires additional 130 bytes per watch on etcd
* It increases kube-apiserver watch CPU usage by 60%
* It increases etcd watch CPU usage by 80%

This is based on testing 1000 watchers, watching a resource with 100 qps (100B objects):
* kube-apiserver CPU usage grew from 3 cores to 5 cores
* etcd CPU usage grew from 1 core to 1.8 cores

It might seem like a big increase, however the presented case is very extreme, that I would not expect such load to happen even in 5k clusters. And in that cases we are running already pretty big machines that 3 code difference is not as significant. cc @wojtek-t @mborsz for opinion.

**Comment by dims** [comment]
cc @chaochn47 @ahrtr @jmhbnz

**Comment by ahrtr** [comment]
> * Overload of the apiserver-etcd watch stream, causing starvation of other watches

Do you mean that even running with an etcd version with the fix to https://github.com/etcd-io/etcd/issues/15402 still may run  into watch starvation due to (1) multiple sub streams sharing the same gRPC stream and (2) one of the sub stream is overloaded? What's the exact etcd version?



> * This opens the question of "why did etcd implement watch by multiplexing multiple watch calls on single watch grpc?". Don't know the answer yet, however it seems like a bad idea to implement it's own stream multiplexing over a protocol that already provides it.

Does this answer your question? : `A single watch stream can multiplex many distinct watches by tagging events with per-watch identifiers. This multiplexing helps reducing the memory footprint and connection overhead on the core etcd cluster.`. Refer to https://etcd.io/docs/v3.5/learning/api/#watch-streams

Also regarding "`a protocol that already provides it`", I am not aware of any such feature being supported by grpc. Not sure whether we are on the same page. Note that the [multiplex supported by gRPC](https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#clients) just means that a [ClientConn](https://pkg.go.dev/google.golang.org/grpc#ClientConn) can be reused/shared by multiple client types, but the [multiplexing supported in etcd](https://etcd.io/docs/v3.5/learning/api/#watch-streams) means that a single physical gprc stream can be shared by multiple logical sub stream/watches. @dfawley does grpc support such feature?




> Etcd client doesn't provide a direct access to managing streams, however client creates a create RPC for unique set of grpc metadata. A separate streams can be forced by adding unique grpc metadata in the watch context.

Yes, it's current implementation. One possible improvement on etcd side is to let clients to forcibly create a separate grpc stream using the [third parameter OpOption](https://github.com/etcd-io/etcd/blob/05aa9796df6a009881e2d13697faf5000915ab74/client/v3/watch.go#L297). It's a backward compatible change, but we need to clearly clarify the use cases.

**Comment by serathius** [comment]
> Does this answer your question? : A single watch stream can multiplex many distinct watches by tagging events with per-watch identifiers. This multiplexing helps reducing the memory footprint and connection overhead on the core etcd cluster.. Refer to https://etcd.io/docs/v3.5/learning/api/#watch-streams

Yes, thanks.

**Comment by serathius** [comment]
> Also regarding "a protocol that already provides it", I am not aware of any such feature being supported by grpc. 

I meant the HTTP2 stream multiplection. Basically you can send multiple concurrent requests and get responses in any order. This can be done by just sending watch RPC per watch request. GRPC will multiplex them underneath. My testing confirms that this comes with overhead, 60%-80% on both client and etcd side. So, there is a tradeoff, by fitting all the watches on single RPC you reduce CPU usage, but all of them will share the same throughput limit.

**Comment by wojtek-t** [comment]
Assuming that the proposed change will be well encapsulated (didn't yet look into the code), I would be supportive for the backport [and I agree that separateing just watchcache seems like a good tradeoff for what we should do]

**Comment by chaochn47** [comment]
Just my guess to make the problem statement of watch starvation more clear before discussing the mitigation and long term fix. s this [channel](https://github.com/etcd-io/etcd/blob/b9eeaf3a18d95df3299015b2ffe3ad4cfa39f800/server/mvcc/watchable_store.go#L112) shared by the 100 watchers buffer being exhausted

**Comment by chaochn47** [comment]
Just my guess to make the problem statement of watch starvation more clear:

Is this [channel](https://github.com/etcd-io/etcd/blob/b9eeaf3a18d95df3299015b2ffe3ad4cfa39f800/server/mvcc/watchable_store.go#L112) shared by the 100 watchers buffer being exhausted due to the rate of consuming this channel is lower by generating data and feeding into the channel (because of gRPC limit)?

@serathius @ahrtr

**Comment by serathius** [comment]
Proposed fix for watch cache starvation issue https://github.com/kubernetes/kubernetes/pull/123532

**Comment by dfawley** [comment]
> @dfawley does grpc support such feature?

No, gRPC does not support any kind of stream multiplexing.  I think the subsequent discussions in this thread are all accurate.  It's perfectly valid to only allow one watch per stream and start a new stream for every watcher, but it's also very normal to support multiple watches on a single stream and demultiplex to the relevant local watcher (in the application).  For example, we do this to implement xDS support, where a single ADS stream handles all subscriptions and the xdsClient internally notifies individual callbacks based upon what resources were updated.

**Comment by chaochn47** [comment]
> So, there is a tradeoff, by fitting all the watches on single RPC you reduce CPU usage, but all of them will share the same throughput limit.

Just curious if anyone has the clue what exact gRPC stream throughput limit we are hitting with thousands of watches multiplexed together? Looks like network problems between client and server is not in the scope of discussion since the reproduce runs locally.

@wojtek-t @liggitt @ahrtr @serathius @dfawley @mborsz

**Comment by serathius** [comment]
We can close as the followup are tracked in separate issues:
* https://github.com/kubernetes/kubernetes/issues/130939
* https://github.com/kubernetes/kubernetes/issues/138264
